### PR TITLE
[SPARK-30763][SQL] Fix java.lang.IndexOutOfBoundsException No group 1 for regexp_extract

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The current implement of `regexp_extract` will throws a unprocessed exception show below:

`SELECT regexp_extract('1a 2b 14m', 'd+')`
```
java.lang.IndexOutOfBoundsException: No group 1
[info] at java.util.regex.Matcher.group(Matcher.java:538)
[info] at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
[info] at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
[info] at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
```
I think should treat this exception well.

### Why are the changes needed?
Fix a bug `java.lang.IndexOutOfBoundsException No group 1 `

### Does this PR introduce any user-facing change?
Yes

### How was this patch tested?
New UT

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/inputs/regexp-function.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+SELECT regexp_extract('1a 2b 14m', '\\d+', 2);
+
+SELECT regexp_extract('1a 2b 14m', '(\\d+)([a-z]+)');
+
+SELECT regexp_extract('1a 2b 14m', '(\\d+)([a-z]+)', 0); 
+
+SELECT regexp_extract('1a 2b 14m', '(\\d+)([a-z]+)', 1); 
+
+SELECT regexp_extract('1a 2b 14m', '(\\d+)([a-z]+)', 2);
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/regexp-function.sql]: Can you move these tests into `sql-tests/inputs/string-functions.sql` insteadf of making a new file?

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/inputs/regexp-function.sql]: Also, can you add tests in `RegexpExpressionsSuite`, too , for the non-codegen path?

Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/inputs/regexp-function.sql]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 2]
```diff
@@ -524,6 +535,8 @@ case class RegExpExtract(subject: Expression, regexp: Expression, idx: Expressio
         $termPattern.matcher($subject.toString());
       if ($matcher.find()) {
         java.util.regex.MatchResult $matchResult = $matcher.toMatchResult();
+        org.apache.spark.sql.catalyst.expressions.RegExpExtract
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: don't hard code the class name. `${classOf[RegExpExtract].getName}. checkGroupIndex`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RegexpExpressionsSuite.scala [hunk 3]
```diff
@@ -293,6 +293,27 @@ class RegexpExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
     val nonNullExpr = RegExpExtract(Literal("100-200"), Literal("(\\d+)-(\\d+)"), Literal(1))
     checkEvaluation(nonNullExpr, "100", row1)
+
+    // invalid group index
+    val row8 = create_row("100-200", "(\\d+)-(\\d+)", 3)
+    val row9 = create_row("100-200", "(\\d+).*", 2)
+    val row10 = create_row("100-200", "\\d+", 1)
+
+    val invalidGroupIndex1 = intercept[AnalysisException] {
+      evaluateWithoutCodegen(expr, row8)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RegexpExpressionsSuite.scala]: As UT we should test both codegen and non-codegen, let's call `checkEvaluation`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: OK

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RegexpExpressionsSuite.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RegexpExpressionsSuite.scala]: Thanks. I know it now.

**Review by maropu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 4]
```diff
@@ -452,6 +452,15 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
   }
 }
 
+object RegExpExtract {
+  def checkGroupIndex(groupCount: Int, groupIndex: Int): Unit = {
+    if (groupCount < groupIndex) {
+      throw new AnalysisException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: We do this check at runtime, so `AnalysisException` doesn't fit. How about `IllegalArgumentException`?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 4 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: OK. Thanks.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
Pending Jenkins.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 4 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Sorry, I missed that...

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@maropu @cloud-fan @dongjoon-hyun Thanks for all your work.

**Comment by gatorsmile** [comment]
@beliefer Does this problem exist in 2.4? If so, submit a backport PR?

**Comment by beliefer** [comment]
@gatorsmile OK

**Comment by gatorsmile** [comment]
It sounds like this is not merged to 3.0. cc @cloud-fan

**Comment by dongjoon-hyun** [comment]
It seems like that. I locally tested and pushed to `branch-3.0`, too.
- https://github.com/apache/spark/commit/9666cf10aac873ad17e95074da5e1a78c8b64b47

**Comment by cloud-fan** [comment]
seems like my merge script got stuck while I was backporting this PR and I forgot to restore, sorry for the trouble!

**Pull request closed without merging**